File: Tutorials/WPF/DevExpress.MVVM/VB/ViewModels/EditOrderViewModel.vb
Imports DevExpress.Mvvm
Imports DevExpress.Xpo
Imports System.Collections
Imports XpoTutorial
Imports System.Linq
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Xpo.DB.Exceptions

Namespace WpfApplicationMvvm.ViewModels
	Public Class EditOrderViewModel
		Inherits ViewModelBase

		Public Property CurrentItem() As Order
			Get
				Return GetProperty(Function() CurrentItem)
			End Get
			Set(ByVal value As Order)
				SetProperty(Function() CurrentItem, value)
			End Set
		End Property
		Private privateUnitOfWork As NestedUnitOfWork
		Protected Property UnitOfWork() As NestedUnitOfWork
			Get
				Return privateUnitOfWork
			End Get
			Private Set(ByVal value As NestedUnitOfWork)
				privateUnitOfWork = value
			End Set
		End Property
		Public Property Customers() As ICollection
			Get
				Return GetProperty(Function() Customers)
			End Get
			Set(ByVal value As ICollection)
				SetProperty(Function() Customers, value)
			End Set
		End Property
		Protected ReadOnly Property MessageBoxService() As IMessageBoxService
			Get
				Return GetService(Of IMessageBoxService)()
			End Get
		End Property
		Protected ReadOnly Property DocumentManagerService() As IDocumentManagerService
			Get
				Return GetService(Of IDocumentManagerService)()
			End Get
		End Property
		Protected Overrides Sub OnParameterChanged(ByVal parameter As Object)
			Dim customer As Customer = TryCast(parameter, Customer)
			Dim order As Order = TryCast(parameter, Order)
			If customer IsNot Nothing Then
				SetCurrentItem(customer)
			ElseIf order IsNot Nothing Then
				SetCurrentItem(order)
			Else
				ResetCurrentItem()
			End If
		End Sub
		Private Sub ResetCurrentItem()
			UnitOfWork = Nothing
			CurrentItem = Nothing
			Customers = Nothing
		End Sub
		Protected Sub SetCurrentItem(ByVal order As Order)
			UnitOfWork = order.Session.BeginNestedUnitOfWork()
			CurrentItem = UnitOfWork.GetNestedObject(order)
			LoadCustomers()
		End Sub
		Protected Sub SetCurrentItem(ByVal customer As Customer)
			UnitOfWork = customer.Session.BeginNestedUnitOfWork()
			CurrentItem = New Order(UnitOfWork)
			CurrentItem.Customer = UnitOfWork.GetNestedObject(customer)
			LoadCustomers()
		End Sub
		Private Async Sub LoadCustomers()
			Customers = Await UnitOfWork.Query(Of Customer)().Select(Function(c) New With {Key c.Oid, Key c.ContactName}).OrderBy(Function(c) c.ContactName).ToListAsync()
		End Sub
		<Command>
		Public Sub Save()
			Try
				UnitOfWork.CommitChanges()
			Catch e1 As LockingException
				MessageBoxService.ShowMessage("The object was modified int another window. Please reload data.", "XPO Tutorial", MessageButton.OK, MessageIcon.Stop)
				Return
			End Try
			DocumentManagerService.ActiveDocument.Close()
		End Sub
		<Command>
		Public Sub Cancel()
			DocumentManagerService.ActiveDocument.Close()
		End Sub
		<Command>
		Public Sub Reload()
			Dim isNewObject As Boolean = UnitOfWork.IsNewObject(CurrentItem)
			If isNewObject Then
				Dim customer As Customer = UnitOfWork.GetParentObject(CurrentItem.Customer)
				SetCurrentItem(customer)
			Else
				Dim order As Order = UnitOfWork.GetParentObject(CurrentItem)
				SetCurrentItem(order)
			End If
		End Sub
	End Class
End Namespace

File: Tutorials/WPF/DevExpress.MVVM/CS/ViewModels/EditOrderViewModel.cs
using DevExpress.Mvvm;
using DevExpress.Xpo;
using System.Collections;
using XpoTutorial;
using System.Linq;
using DevExpress.Mvvm.DataAnnotations;
using DevExpress.Xpo.DB.Exceptions;

namespace WpfApplicationMvvm.ViewModels {
    public class EditOrderViewModel :ViewModelBase {
        public Order CurrentItem {
            get { return GetProperty(() => CurrentItem); }
            set { SetProperty(() => CurrentItem, value); }
        }
        protected NestedUnitOfWork UnitOfWork { get; private set; }
        public ICollection Customers {
            get { return GetProperty(() => Customers); }
            set { SetProperty(() => Customers, value); }
        }
        protected IMessageBoxService MessageBoxService {
            get { return GetService<IMessageBoxService>(); }
        }
        protected IDocumentManagerService DocumentManagerService {
            get { return GetService<IDocumentManagerService>(); }
        }
        protected override void OnParameterChanged(object parameter) {
            Customer customer = parameter as Customer;
            Order order = parameter as Order;
            if (customer != null)
                SetCurrentItem(customer);
            else if (order != null)
                SetCurrentItem(order);
            else ResetCurrentItem();
        }
        private void ResetCurrentItem() {
            UnitOfWork = null;
            CurrentItem = null;
            Customers = null;
        }
        protected void SetCurrentItem(Order order) {
            UnitOfWork = order.Session.BeginNestedUnitOfWork();
            CurrentItem = UnitOfWork.GetNestedObject(order);
            LoadCustomers();
        }
        protected void SetCurrentItem(Customer customer) {
            UnitOfWork = customer.Session.BeginNestedUnitOfWork();
            CurrentItem = new Order(UnitOfWork);
            CurrentItem.Customer = UnitOfWork.GetNestedObject(customer);
            LoadCustomers();
        }
        private async void LoadCustomers() {
            Customers = await UnitOfWork.Query<Customer>()
                .Select(c => new { c.Oid, c.ContactName })
                .OrderBy(c => c.ContactName)
                .ToListAsync();
        }
        [Command]
        public void Save() {
            try {
                UnitOfWork.CommitChanges();
            } catch (LockingException) {
                MessageBoxService.ShowMessage("The object was modified int another window. Please reload data.", "XPO Tutorial", MessageButton.OK, MessageIcon.Stop);
                return;
            }
            DocumentManagerService.ActiveDocument.Close();
        }
        [Command]
        public void Cancel() {
            DocumentManagerService.ActiveDocument.Close();
        }
        [Command]
        public void Reload() {
            bool isNewObject = UnitOfWork.IsNewObject(CurrentItem);
            if (isNewObject) {
                Customer customer = UnitOfWork.GetParentObject(CurrentItem.Customer);
                SetCurrentItem(customer);
            } else {
                Order order = UnitOfWork.GetParentObject(CurrentItem);
                SetCurrentItem(order);
            }
        }
    }
}
